### PR TITLE
Improved error reporting in inbox approve REST resource

### LINKED ISSUE #6129: InboxResource error reporting insufficient
Currently, if there is an error approving an inbox entry, there is no logging as to the cause of the error, and the resource returns 404 "Thing not found in inbox". This is not always the case and it makes it rather difficult to debug without stepping into the code.
At least one other issue that has come up is reported as "Duplicate channels", but I see that the inbox will throw different reasons.

I suggest to return the cause in the 404, and also for PaperUI to display the text from the return and not just the 404. Additionally, the exception should be logged.

### PR DESCRIPTION
**Pull request by cdjackson**
As per #6129 it's currently not possible to work out why an inbox entry failed to be approved - the current reason is "Inbox entry not found" and this is not the only reason. This PR returns the reason for failing to approve an inbox entry to the UI, and also logs the error in the logfile.

Partially fixes #6129 (PaperUI probably should be updated to report the error as 404 is insufficient to allow a user to know what happened).

Log entry will now look like this -:
```
Thing zwave:device:15f24d360d4:node18 unable to be approved: Duplicate channels zwave:device:15f24d360d4:node18:alarm_heat
```

Signed-off-by: Chris Jackson <chris@cd-jackson.com>

### CODE REVIEW

File: bundles/io/org.eclipse.smarthome.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/internal/discovery/InboxResource.java [hunk 1]
```diff
@@ -100,7 +103,9 @@ public Response approve(@HeaderParam(HttpHeaders.ACCEPT_LANGUAGE) @ApiParam(valu
         try {
             thing = inbox.approve(thingUIDObject, notEmptyLabel);
         } catch (IllegalArgumentException e) {
-            return JSONResponse.createErrorResponse(Status.NOT_FOUND, "Thing not found in inbox");
+            logger.error("Thing {} unable to be approved: {}", thingUID, e.getLocalizedMessage());
+            return JSONResponse.createErrorResponse(Status.NOT_FOUND,
+                    "Thing unable to be approved: " + e.getLocalizedMessage());
```
Comment (maggu2810) [inline bundles/io/org.eclipse.smarthome.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/internal/discovery/InboxResource.java]: I agree with the log message but adding the message of an exception to the REST response seems wrong to me. Exceptions and its messages are part of the internals and should not be exposed to the caller of a REST API.
Or does the inbox approve methods state that the exception message (if IAE) is exposed to external callers and this message must not contain any internal information?

**Comment by cdjackson** [comment]
Normally, I would agree, however if you look at the implementation, it is throwing use friendly messages. At the moment, the message (inbox entry doesn’t exist) is incorrect, and the information provided in the exception is perfectly user friendly.

Directly in the approve method, we have
            throw new IllegalArgumentException("Thing UID must not be null");
            throw new IllegalArgumentException("No Thing with UID " + thingUID.getAsString() + " in inbox");

Additionally, elsewhere we get things like I showed in the original PR relating to duplicate channels.

I don’t overly mind if we don’t report it to the user, although I think at the moment they are in the dark, and as someone has gone to the effort of providing user friendly messages in all these exceptions, it seems a waste not to use them ;)

**Comment by maggu2810** [comment]
@cdjackson Giving more information to the caller of the REST resource about what is going wrong is okay and should be implemented.

But changing the REST implementation to forward the messages of the excpetion that has been generated by another implementation class seems wrong.
Can't we use e.g. different exceptions in the approve method that the REST resource can differ?

I don't think that in the next years if someone changes the inbox implementation we remember all the time that the message used in exceptions of the approve method is forwared to the caller of the REST API.
I also don't think implementations should be required to use exception messages that can be forwared to external callers. Where do we state such requirements?

**Comment by cdjackson** [comment]
Ok, I’ll revert this then - at the moment, the message provided to the user is completely useless, (just 404 - nothing else. If you look at the detail, it says the inbox entry doesn’t exist) so I think it’s worth someone looking at it at some stage.

File: bundles/io/org.eclipse.smarthome.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/internal/discovery/InboxResource.java [hunk 2]
```diff
@@ -100,7 +103,7 @@ public Response approve(@HeaderParam(HttpHeaders.ACCEPT_LANGUAGE) @ApiParam(valu
         try {
             thing = inbox.approve(thingUIDObject, notEmptyLabel);
         } catch (IllegalArgumentException e) {
-            return JSONResponse.createErrorResponse(Status.NOT_FOUND, "Thing not found in inbox");
+            logger.error("Thing {} unable to be approved: {}", thingUID, e.getLocalizedMessage());
```
Comment (maggu2810) [inline bundles/io/org.eclipse.smarthome.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/internal/discovery/InboxResource.java]: Hm, don't we need still a return of "404"?

IMHO you should keep add

```
return JSONResponse.createErrorResponse(Status.NOT_FOUND, "Thing unable to be approved");
```

after the log mesage.

Comment (cdjackson) [reply bundles/io/org.eclipse.smarthome.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/internal/discovery/InboxResource.java]: Hmm - strange. I thought this was still there, but I guess the revert button removed it. I'll add it back - it was not intentional to remove it.

**Pull request merged**